src/lib/device-online-state.ts

```diff
@@ -0,0 +1,320 @@
+	REDIS_PORT,
+	API_HEARTBEAT_STATE_TIMEOUT_SECONDS,
+} from './config';
+
+export const getPollIntervalForDevice = (uuid: string): Bluebird<number> => {
+	return resinApi
+		.get({
+			resource: 'device_config_variable',
+			passthrough: {
+				req: root,
+			},
+			options: {
+				$select: 'value',
+				$expand: {
+					device: {
+						$filter: { uuid },
+					},
+				},
+				$filter: {
+					name: {
+						$in: [
+							'BALENA_SUPERVISOR_POLL_INTERVAL',
+							'RESIN_SUPERVISOR_POLL_INTERVAL',
+						],
+					},
+				},
+			},
+		})
+		.then(([pollInterval]: Array<{ value: number }>) => {
+			if (pollInterval) {
+				return pollInterval.value;
+			}
+
+			return resinApi
+				.get({
+					resource: 'application_config_variable',
+					passthrough: {
+						req: root,
+					},
+					options: {
+						$select: 'value',
+						$top: 1,
+						$filter: {
+							application: {
+								$any: {
+									$alias: 'a',
+									$expr: {
+										a: {
+											owns__device: {
+												$any: {
+													$alias: 'd',
+													$expr: {
+														d: {
+															uuid,
+														},
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+							name: {
+								$in: [
+									'BALENA_SUPERVISOR_POLL_INTERVAL',
+									'RESIN_SUPERVISOR_POLL_INTERVAL',
+								],
+							},
+						},
+					},
+				})
+				.then(([pollInterval]: Array<{ value: number }>) => {
+					if (pollInterval) {
+						return pollInterval.value;
+					}
+
+					return DEFAULT_SUPERVISOR_POLL_INTERVAL;
+				});
+		})
+		.then(pollInterval => {
+			return Math.max(pollInterval, DEFAULT_SUPERVISOR_POLL_INTERVAL);
+		});
+};
+
+export enum DeviceOnlineStates {
+	Unknown = -2,
+	Timeout = -1,
+	Offline = 0,
+	Online = 1,
+}
+
+class DeviceOnlineStateManager {
+	private static readonly REDIS_NAMESPACE = 'device-online-state';
+	private static readonly EXPIRED_QUEUE = 'expired';
+	private static readonly RSMQ_READ_TIMEOUT = 30;
+
+	rsmq: RedisSMQ;
+	redis: PromisifedRedisClient;
+
+	constructor() {
+		// create a new Redis client...
+		this.redis = createPromisifedRedisClient({
+			host: REDIS_HOST,
+			port: REDIS_PORT,
+		});
+
+		// initialise the RedisSMQ object using our Redis client...
+		this.rsmq = new RedisSMQ({
+			client: this.redis,
+			ns: DeviceOnlineStateManager.REDIS_NAMESPACE,
+		});
+
+		// create the RedisMQ queue and start consuming messages...
+		this.rsmq
+			.createQueueAsync({ qname: DeviceOnlineStateManager.EXPIRED_QUEUE })
+			.catch(err => {
+				if (err.name !== 'queueExists') {
+					throw err;
+				}
+			})
+			.then(() => {
+				return this.consume();
+			});
+	}
+
+	private updateDeviceModel(uuid: string, newState: DeviceOnlineStates) {
+		// patch the api_heartbeat_state value to the new state...
+		const body = {
+			api_heartbeat_state: newState,
+		};
+
+		return resinApi
+			.patch({
+				resource: 'device',
+				options: {
+					$filter: {
+						uuid,
+						$not: body,
```

Comment:
This limits the DB write scope to only update values which are currently different. It is pointless to lock rows for update when nothing with be changed.